README.md
# TLE module for Yii2

Module extension provides [TLE](https://en.wikipedia.org/wiki/Two-line_element_set) download and storage component
for Yii2. It relies on [Space Track API](https://www.space-track.org/documentation#/api) to get relevant TLE data.

## Requirements

This module relies on [CURL PHP extension](http://php.net/manual/book.curl.php).

## Installation

In order to use the module:

1. Add it to `composer.json` file of your Yii2 application:

        "solarpatrol/yii2-tle": "^1.0.0"

    or run:
    
        $ composer install solarpatrol/yii2-tle

2. Add module to your application's [configuration](http://www.yiiframework.com/doc-2.0/guide-concept-configurations.html)
and set up `storage` component:

        'modules' => [
            'tle' => [
                'class' => 'solarpatrol\tle\Module',
                'components' => [
                    'storage' => [
                        'class' => 'solarpatrol\tle\FileStorage',
                        'actualDaysCount' => 10,
                        'storagePath' => '@runtime/tle',
                        'spaceTrackLogin' => 'myspace',
                        'spaceTrackPassword' => 'passw0rd',
                        'dirMode' => 0777,
                        'fileMode' => 0666,
                        ...
                    ]
                ]
            ]
            ...
        ]

    Refer to [configuration](#configuration) section in order to get how to configure `storage` component.
    
3. In case of using modules' web and console controllers and actions (when direct access to the module
`\Yii::$app->getModule('tle')` is not planned) add the module to bootstrapping stage    

4. If `solarpatrol\tle\DatabaseStorage` is used as `storage` component then apply module migrations to create a table
to store TLEs in:

        ./yii migrate --migrationPath=@vendor/solarpatrol/yii2-tle/migrations
        
## Usage        

Access TLE storage component:

    $storage = \Yii::$app->getModule('tle')->storage;

Download actual TLEs for Terra, Aqua and Meteor-M №2 satellites in a range of 3 days and add them to the storage
(satellites are identified by their [NORAD ids](https://en.wikipedia.org/wiki/Satellite_Catalog_Number)):

    $storage->update([25994, 27424, 40069], time() - 86400 * 3, time());
    
Add TLE for Terra to the storage manually (not recommended):

    $storage->add([
        '0 25994',
        '1 40069U 14037A   16200.39183603 -.00000022  00000-0  99041-5 0  9999',
        '2 40069  98.7043 255.1534 0006745  96.2107 263.9838 14.20632312105160'
    ]);
    
Get all TLEs for Terra in the storage within specified time range:
    
    $tles = $storage->get(25994, '2016-07-18', '2016-07-20T23:59:59');
    
Get all TLEs for Terra and Aqua in the storage within specified time range (result is an associative array where
keys are NORAD identifiers of satellites and values are arrays of found TLEs):

    $tles = $storage->get([25994, 27424], '2016-07-18', '2016-07-20T23:59:59');
    
Find closest actual TLE for Terra within specified time range:

    $tles = $storage->get(25994, '2016-07-18', '2016-07-20T23:59:59');
    $tle = Storage::getClosest($tles, '2016-07-19T16:44:44');
    
Remove Terra TLE from the storage manually (not recommended):

    $storage->remove([
        '0 25994',
        '1 40069U 14037A   16200.39183603 -.00000022  00000-0  99041-5 0  9999',
        '2 40069  98.7043 255.1534 0006745  96.2107 263.9838 14.20632312105160'
    ]);
     
Get a list of all known launched satellites:
     
    $satellites = $storage->getSatcat();
    foreach($satellites as $satellite) {
        print $satellite['id'] . ' ' . $satellite['name'] . "\n";
    }

## Configuration

Two implementations of `solarpatrol\tle\Storage` component are supported:

- `solarpatrol\tle\FileStorage` — stores TLEs in file system;
- `solarpatrol\tle\DatabaseStorage` — stores TLEs in database.

Both of them have the following common sensitive configuration properties:

- `actualDaysCount` — number of days to use to perform requested action if both start and end of time range are
not specified (defaults to `5`);
- `spaceTrackLogin` — [Space Track](https://www.space-track.org/) account's name (e-mail) that can be created
[here](https://www.space-track.org/auth/createAccount) (required);
- `spaceTrackPassword` — password for Space Track account (required);
- `connectionTimeout` — timeout in seconds for CURL requests to Space Track API (defaults to `30`);
- `enableCaching` — whether CURL requests' results should be cached (defaults to `true`);
- `cacheExpiration` — cache expiration in seconds (defaults to `21600`);
- `userAgent` — user agent header to set for CURL requests to Space Track API;
- `proxyHost` — proxy host (if used);
- `proxyPort` — proxy port (if used);
- `proxyAuthLogin` — proxy authentication login (if proxy requires authentication);
- `proxyAuthPassword` — proxy authentication password (if proxy requires authentication).

`FileStorage` has the following additional properties:

- `storagePath` — path to directory to store TLE files in (defaults to `@runtime/tle`);
- `dirMode` — directories' creation mode (defaults to `0775`);
- `fileMode` — files' creation mode.

`DatabaseStorage` has the following additional properties:

- `db` — name, configuration array or instance of database connection used to access TLE table (defaults to `db`).

## Web requests

In order to use built-in web controller specify the module in [bootstrap](http://www.yiiframework.com/doc-2.0/guide-structure-modules.html#bootstrapping-modules)
configuration of web application:

    'bootstrap' => ['tle', ...other modules],
    'modules' => [
        'tle' => [
            'components' => [
                'storage' => [
                    'class' => 'solarpatrol\tle\FileStorage',
                    ...
                ]
            ]
        ]
    ],
    ...
    
and run web action:

    /tle/request?id[]=25994&id[]=27424&startTime=2016-09-21&endTime=2016-09-23T23:59:59
     
If you want to change controller's identifier (which is `tle` by default) set `webControllerId`:
   
    'module' => 
        'class' => 'solarpatrol\tle\FileStorage',
        'webControllerId` => 'orbit`
    ]
     
Run web action:
     
    /orbit/request?id[]=25994&id[]=27424&startTime=2016-09-21&endTime=2016-09-23T23:59:59
    
The following web actions are available:

1. Request TLEs for given satellites within given time range:

        /tle/request
    
    - `id` is array of satellites' NORAD identifiers;
    - `startTime` is start of time range in ISO 8601 or Unix timestamp (optional, if omitted then a moment
    `actualDaysCount` days earlier than `endTime` is taken);
    - `endTime` is end of time range in ISO 8601 or Unix timestamp (optional, if omitted then current system time is
    taken);
    - `download` if it's set then attempt to download TLEs missed in the storage will be executed. 
    
    ### Examples
    
    - Request TLEs for Terra and Aqua for recent five days:
        
            /tle/request?id[]=25994&id[]=27424
        
    - Request TLEs for Terra, Aqua and Meteor-M №2 in time range 21st of September, 2016 — 23rd of September,
    2016:

            /tle/request?id[]=25994&id[]=27424id[]=40069&startTime=2016-09-21&endTime=2016-09-23T23:59:59

## Console commands

In order to use console commands specify the module in [bootstrap](http://www.yiiframework.com/doc-2.0/guide-structure-modules.html#bootstrapping-modules)
configuration of console application:

    'bootstrap' => ['tle', ...other modules],
    'modules' => [
        'tle' => [
            'components' => [
                'storage' => [
                    'class' => 'solarpatrol\tle\FileStorage',
                    ...
                ]
            ]
        ]
    ],
    ...
    
and run a command:
    
    ./yii tle/update 25994
    
If you want to change controller's identifier (which is `tle` by default) set `consoleControllerId`:
  
    'module' => 
        'class' => 'solarpatrol\tle\FileStorage',
        'consoleControllerId` => 'orbit`
    ]
    
Run a command:
    
    ./yii orbit/update 25994
    
The following console commands are available:
    
1. Download TLEs and save them in the storage (default): 
    
        ./yii tle/update [ids] [--existing] [--all] [--startTime] [--endTime]
    
    where

    - `ids` is a set of NORAD identifiers (used when both `existing` and `all` arguments are missed);
    - `existing` determines `ids` automatically basing on TLEs already existing in the storage;
    - `all` determines `ids` automatically basing on Satellite Catalog of all known launched satellites;
    - `startTime` is start of time range in ISO 8601 or Unix timestamp (optional, if omitted then a moment
    `actualDaysCount` days earlier than `endTime` is taken);
    - `endTime` is end of time range in ISO 8601 or Unix timestamp (optional, if omitted then current system time is
    taken).
    
    ### Examples
    
    - Download and store TLEs for Terra and Aqua for recent five days:
    
            ./yii tle/update 25994 27424
        
    - Download and store TLEs for Terra, Aqua and Meteor-M №2 in time range 21st of September, 2016 — 23rd of September,
    2016:
    
            ./yii tle/update 25994 27424 40069 --startTime=2016-09-21 --endTime=2016-09-23T23:59:59
            
    - Download TLEs for all satellites which are already in the storage:
    
            ./yii tle/update --existing
            
    - Download TLEs for all known launched satellites:
    
            ./yii tle/update --all

## Contribution

Before making a pull request, please, be sure that your changes are rebased to `dev` branch.


CHANGES.md
## Change Log

### 1.3.0

- Option to force missed TLEs download in HTTP request is supported.

### 1.2.0

- [Satcat](https://www.space-track.org/basicspacedata/modeldef/class/satcat/format/html) retrieval is supported.
- Options to determine `ids` automatically are supported in update console command.  

### 1.1.0

- API is changed.

### 1.0.0

- Initial release.


TleStorage.php
<?php

namespace app\components\tle;

use yii\base\Component;
use yii\base\Exception;
use yii\helpers\Json;

abstract class TleStorage extends Component
{
    const TLE_LINE_LENGTH = 69;
    const SECONDS_PER_DAY = 86400;

    public $spaceTrackUrl = 'https://www.space-track.org';
    public $spaceTrackLogin = '';
    public $spaceTrackPassword = '';
    public $connectionTimeout = 30;
    public $cookiePath = 'space-track-cookie';
    public $proxyHost;
    public $proxyPort;
    public $proxyAuthLogin = false;
    public $proxyAuthPassword = '';
    public $userAgent;
    public $enableCaching = true;
    public $cacheExpiration = 10800;

    protected $defaultCurlOptions = [];

    public function init()
    {
        parent::init();

        $this->defaultCurlOptions = [
            CURLOPT_URL => $this->spaceTrackUrl,
            CURLOPT_HTTPHEADER => [
                'Accept: */*',
                'Connection: Keep-Alive'
            ],
            CURLOPT_HEADER => true,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_FOLLOWLOCATION => true,
            CURLOPT_AUTOREFERER => true,
            CURLOPT_TIMEOUT => $this->connectionTimeout,
            CURLOPT_CONNECTTIMEOUT => $this->connectionTimeout,
            CURLOPT_SSL_VERIFYHOST, true,
            CURLOPT_SSL_VERIFYPEER, true
        ];

        if (isset($this->proxyHost) && isset($this->proxyPort)) {
            $this->setProxy($this->proxyHost, $this->proxyPort, $this->proxyAuthLogin, $this->proxyAuthPassword);
        }

        if (isset($this->userAgent)) {
            $this->defaultCurlOptions[CURLOPT_USERAGENT] = $this->userAgent;
        }
    }

    /**
     * @param int $id
     * @param int $timestamp
     * @return bool
     */
    abstract function exists($id, $timestamp);

    /**
     * @param int $id
     * @param string $line1
     * @param string $line2
     * @return bool
     */
    abstract function add($id, $line1, $line2);

    /**
     * @param int $id
     * @param int $startTimestamp
     * @param int $endTimestamp
     * @return mixed
     */
    abstract function getRange($id, $startTimestamp, $endTimestamp);

    /**
     * @param int $id
     * @param int $timestamp
     * @return bool
     */
    abstract function remove($id, $timestamp);

    public function get($id, $timestamp, $actualDaysCount = 5)
    {
        $startTime = $timestamp - self::SECONDS_PER_DAY * $actualDaysCount;
        $endTime = $timestamp + self::SECONDS_PER_DAY * $actualDaysCount;

        $tles = $this->getRange($id, $startTime, $endTime);

        $closestTle = null;

        foreach ($tles as $tle) {
            if ($closestTle === null || abs($timestamp - $tle['timestamp']) < abs($timestamp - $closestTle['timestamp'])) {
                $closestTle = $tle;
                continue;
            }
            break;
        }

        return $closestTle;
    }

    public function update(array $ids, $timestamp, $actualDaysCount = 5)
    {
        $year = intval(gmdate('Y', $timestamp));
        $month = intval(gmdate('n', $timestamp));
        $day = intval(gmdate('j', $timestamp));
        $currentDayStartTimestamp = gmmktime(0, 0, 0, $month, $day, $year);

        $startTimestamp = $currentDayStartTimestamp - $actualDaysCount * TleStorage::SECONDS_PER_DAY;
        $endTimestamp = $currentDayStartTimestamp + ($actualDaysCount + 1) * TleStorage::SECONDS_PER_DAY;

        return $this->updateRange($ids, $startTimestamp, $endTimestamp);
    }

    public function updateRange(array $ids, $startTimestamp, $endTimestamp)
    {
        $data = $this->download($ids, $startTimestamp, $endTimestamp);
        if (!$data) {
            return false;
        }

        foreach ($data as $dataItem) {
            if ($this->exists($dataItem['id'], $dataItem['epochTimestamp'])) {
                continue;
            }

            if (!$this->add($dataItem['id'], $dataItem['line1'], $dataItem['line2'])) {
                return false;
            }
        }

        return true;
    }

    public function download(array $ids, $startTimestamp, $endTimestamp)
    {
        $epochRange = gmdate('Y-m-d', $startTimestamp) . '--' . gmdate('Y-m-d', $endTimestamp);

        $url = sprintf('%s/basicspacedata/query/class/tle/format/json/EPOCH/%s/NORAD_CAT_ID/%s/orderby/EPOCH%%20desc',
            $this->spaceTrackUrl, $epochRange, implode(',', $ids)
        );

        $cacheKey = 'tle-download:' . sha1(serialize([
            'epochStartTime' => $startTimestamp,
            'epochEndTime' => $endTimestamp,
            'ids' => $ids
        ]));

        if ($this->enableCaching) {
            $data = \Yii::$app->cache->get($cacheKey);
            if ($data !== false) {
                return $data;
            }
        }

        $cookiePath = null;
        try {
            $loginResult = $this->login();
            if ($loginResult['response']['http_code'] !== 200) {
                return false;
            }

            $cookiePath = $loginResult['cookiePath'];

            $fetchResult = $this->curl([
                CURLOPT_URL => $url,
                CURLOPT_HEADER => false,
                CURLOPT_COOKIEFILE => $cookiePath
            ]);
            if ($fetchResult['response']['http_code'] !== 200 ||
                $fetchResult['response']['content_type'] !== 'application/json'
            ) {
                $this->logout($cookiePath);
                return false;
            }

            $this->logout($cookiePath);

            $data = $this->parseTleData($fetchResult['output']);
            if ($this->enableCaching) {
                \Yii::$app->cache->set($cacheKey, $data, $this->cacheExpiration);
            }

            return $data;
        }
        catch (Exception $e) {
            if ($cookiePath !== null) {
                $this->logout($cookiePath);
            }
            throw $e;
        }
    }

    protected function parseTleData(&$output)
    {
        $outputData = Json::decode($output);

        $data = [];
        foreach ($outputData as $outputDataItem) {
            $id = intval($outputDataItem['NORAD_CAT_ID']);
            $line1 = substr($outputDataItem['TLE_LINE1'], 0, self::TLE_LINE_LENGTH);
            $line2 = substr($outputDataItem['TLE_LINE2'], 0, self::TLE_LINE_LENGTH);
            $epoch = $outputDataItem['EPOCH'];
            $epochParsed = date_parse_from_format('Y.m.d H:i:s', $outputDataItem['EPOCH']);

            $data[] = [
                'id' => $id,
                'line1' => $line1,
                'line2' => $line2,
                'epochTime' => $epoch,
                'epochTimestamp' => $this->getEpochTimestamp($line1),
                'epochTimestampParsed' => gmmktime(
                    $epochParsed['hour'],
                    $epochParsed['minute'],
                    $epochParsed['second'],
                    $epochParsed['month'],
                    $epochParsed['day'],
                    $epochParsed['year']
                )
            ];
        }

        return $data;
    }

    protected function getEpochTimestamp($line1)
    {
        $year = intval(substr($line1, 18, 2));
        $year = ($year >= 57 ? 1900 : 2000) + $year;
        $seconds = intval((floatval(substr($line1, 20, 12)) - 1) * 86400);
        return gmmktime(0, 0, 0, 1, 1, $year) + $seconds;
    }

    protected function curl($curlOptions = array(), $options = array())
    {
        foreach ($this->defaultCurlOptions as $key => $value) {
            if (!isset($curlOptions[$key])) {
                $curlOptions[$key] = $value;
            }
        }

        $curlHandler = isset($options['curlHandler']) ? $options['curlHandler'] : curl_init();
        foreach ($curlOptions as $curlOptionName => $curlOptionValue) {
            curl_setopt($curlHandler, $curlOptionName, $curlOptionValue);
        }

        $output = curl_exec($curlHandler);
        $response = curl_getinfo($curlHandler);

        $result = array(
            'response' => $response,
            'output' => $output
        );

        if ($output === false) {
            $result['errorCode'] = curl_errno($curlHandler);
            $result['errorMessage'] = curl_error($curlHandler);
        }

        if (isset($options['closeCurlHandler']) && $options['closeCurlHandler'] === false) {
            $result['curlHandler'] = $curlHandler;
        }
        else {
            curl_close($curlHandler);
        }
        return $result;
    }

    protected function login()
    {
        $cookiePath = tempnam(sys_get_temp_dir(), $this->cookiePath);
        $result = $this->curl([
            CURLOPT_URL => $this->spaceTrackUrl . '/ajaxauth/login',
            CURLOPT_POST => true,
            CURLOPT_POSTFIELDS => [
                'identity' => $this->spaceTrackLogin,
                'password' => $this->spaceTrackPassword
            ],
            CURLOPT_COOKIEJAR => $cookiePath
        ]);

        $result['cookiePath'] = $cookiePath;
        return $result;
    }

    protected function logout($cookiePath)
    {
        return $this->curl([
            CURLOPT_URL => $this->spaceTrackUrl . '/ajaxauth/logout',
            CURLOPT_POST => false,
            CURLOPT_COOKIEFILE => $cookiePath
        ]);
    }

    public function setProxy($host, $port = 0, $authLogin = false, $authPassword = '')
    {
        if (!$host) {
            $this->defaultCurlOptions[CURLOPT_PROXY] = null;
            return;
        }

        $proxy = $host;
        if ($port) {
            $proxy .= ':' . $port;
        }
        $this->defaultCurlOptions[CURLOPT_PROXY] = $proxy;

        if ($authLogin) {
            $this->defaultCurlOptions[CURLOPT_PROXYUSERPWD] = $authLogin . ':' . $authPassword;
        }
    }
}